src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/CastHelpers.cs

```diff
@@ -517,6 +517,45 @@ private static void ArrayTypeCheck_Helper(object obj, void* elementType)
             }
         }
 
+        // Helpers for boxing
+        [DebuggerHidden]
+        internal static object? Box_Nullable(MethodTable* srcMT, ref byte nullableData)
+        {
+            Debug.Assert(srcMT->IsNullable);
+
+            // If 'hasValue' is false, return null.
+            if (!Unsafe.As<byte, bool>(ref nullableData))
+                return null;
+
+            // Allocate a new instance of the T in Nullable<T>.
+            MethodTable* dstMT = srcMT->InstantiationArg0();
+            ref byte srcValue = ref Unsafe.Add(ref nullableData, srcMT->NullableValueAddrOffset);
+
+            // Delegate to non-nullable boxing implementation
+            return Box(dstMT, ref srcValue);
+        }
+
+        [DebuggerHidden]
+        internal static object Box(MethodTable* typeMT, ref byte unboxedData)
+        {
+            Debug.Assert(typeMT != null);
+            Debug.Assert(typeMT->IsValueType);
+
+            object boxed = RuntimeTypeHandle.GetThreadLocalAllocInternal(typeMT, out uint bytes) ??
```

Comment:
I think it would be better to have this fast path in `InternalAllocNoChecks` so that the other caller of `InternalAllocNoChecks` benefits from it too. I am fine with losing the bit of perf from passing the size via `out` argument.

Reply:
Updated, running benchmark: https://github.com/EgorBot/runtime-utils/issues/348#issuecomment-2845059677

Reply:
The impact, compared to FCall, seems significant.

Reply:
I think the ask was opposite of what you did... ie. remove the `out` parameter, keep `GetThreadLocalAllocInternal` FCall, just move the `RuntimeTypeHandle.GetThreadLocalAllocInternal(...) ?? InternalAllocNoChecks(...)` expression to some `InternalAllocNoChecks` managed method.